mmv1/third_party/terraform/resources/resource_storage_bucket_object.go

```diff
@@ -161,16 +163,37 @@ func resourceStorageBucketObject() *schema.Resource {
 				DiffSuppressFunc: compareCryptoKeyVersions,
 				Description:      `Resource name of the Cloud KMS key that will be used to encrypt the object. Overrides the object metadata's kmsKeyName value, if any.`,
 			},
+
+			"customer_encryption": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+				Sensitive:   true,
+				Description: `Base64 encoded customer supplied encryption key.`,
+				ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
+					decodedKey, err := base64.StdEncoding.DecodeString(val.(string))
+					if err != nil {
+						errs = append(errs, fmt.Errorf("Failed to decode (base64) customer_encryption, expecting valid base64 encoded key"))
+					}
+					if len(decodedKey) != 32 {
+						errs = append(errs, fmt.Errorf("Wrong customer_encryption key provided: not a 32-byte AES-256 key"))
+					}
+					return
+				},
+			},
```

Comment:
can we model this off of the `ObjectCustomerEncryption`

https://github.com/googleapis/google-api-go-client/blob/master/storage/v1/storage-gen.go#L1846

```
			"customer_encryption": {
				Type:        schema.TypeList,
				MaxItems:    1,
				Optional:    true,
				Sensitive: true,
				Description: `SHA256 hash value of the encryption key; encoded using base64.`,
				Elem: &schema.Resource{
					Schema: map[string]*schema.Schema{
						"encryption_algorithm": {
							Type:         schema.TypeString,
							Required: 	  true,
							ForceNew:     true
							Description:  `The encryption algorithm.`,
						},
						"key_sha256": {
							Type:         schema.TypeString,
							Required: 	  true,
							ForceNew:     true
							Description:  `The encryption algorithm.`,
							ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
								decodedKey, err := base64.StdEncoding.DecodeString(val.(string)
								if err != nil {
									errs = append(errs, fmt.Errorf("Failed to decode (base64) customer_encryption, expecting valid base64 encoded key"))
								}
								return
							},
						},
					},
				},
```

Reply:
I did this but with attributes 'encryption_algorithm' and 'encryption_key' (it's sha256 is calculated automatically and added to headers)

Reply:
makes sense !